CORDA-3329 Exceptions thrown in raw vault observers can cause critical issues

kyriathar opened the pull request:
Observers registered on NodeVaultService#rawUpdates, if they throw an exception when called from serviceHub#recordTransactions and if this exception is not handled by the flow hospital, then this leads to the transaction not being recorded in the local vault. This could get the ledger in an out of sync state.

In the specific case this happens within FinalityFlow#notariseAndRecord this leads to the transaction being notarized but not recorded in the local vault nor broadcasted in any counter party. The -failed to be recorded locally- transaction and its output states are not visible to any vault, and its input states not able to consumed by a new transaction, since they are recorded as consumed within the Notary. In this specific case we need not loose, by any means, the current transaction.

We will handle all cases by catching all exceptions thrown from serviceHub#recordTransactions, wrapping them with a HospitalizeFlowException and throwing it instead. The flow will get to the hospital for observation to be retried from previous checkpoint on next node restart.

kyriathar-r3 committed:
Wrapping all exceptions thrown from serviceHub#recordTransactions, in FinalityFlow#notariseAndRecord with a HospitalizeFlowException,
throwing it instead.

Tests added to test that transactions that get successfully notarised but fail to record locally get hospitalised.

blsemo commented on core-tests/src/test/kotlin/net/corda/coretests/flows/FinalityFlowTests.kt:
```diff
@@ -83,6 +94,112 @@ class FinalityFlowTests : WithFinality {
+        val future = aliceNode.services.startFlow(CashPaymentFlow(1000.POUNDS, bobNode.info.singleIdentity())).resultFuture
+        mockNet.runNetwork()
+
+        assertFailsWith<TimeoutException> {
```
Wouldn't it be better to wait for a future you can set in the onFlowKeptForOvernightObservation observer? Thus, the test would not hang for 5 seconds before carrying on.
Also, waiting for a time out exception is potentially flaky. What if something is a bit slow on the build server and the flow genuinely takes longer than 5 seconds?

blsemo commented on core-tests/src/test/kotlin/net/corda/coretests/flows/FinalityFlowTests.kt:
```diff
@@ -83,6 +94,112 @@ class FinalityFlowTests : WithFinality {
+        val future = aliceNode.services.startFlow(CashPaymentFlow(1000.POUNDS, bobNode.info.singleIdentity())).resultFuture
+        mockNet.runNetwork()
+
+        assertFailsWith<TimeoutException> {
```
Same as above. You could also check there what exception you actually got.

blsemo commented on core-tests/src/test/kotlin/net/corda/coretests/flows/FinalityFlowTests.kt:
```diff
@@ -83,6 +94,112 @@ class FinalityFlowTests : WithFinality {
+
+        aliceNode.services.vaultService.rawUpdates.subscribe(
+            { throw Exception("Error in Observer#onNext") },
+            { throw Exception("Error in Observer#onError") }) // register a -throwing Exception- Observer.
```
What does adding this `onError` handler add to the test case?

kyriathar replied:
Exception thrown from observer will no longer be OnErrorNotImplementedException , but OnErrorFailedException.

blsemo commented on core-tests/src/test/kotlin/net/corda/coretests/flows/FinalityFlowTests.kt:
```diff
@@ -83,6 +94,112 @@ class FinalityFlowTests : WithFinality {
+        }
+        assertEquals(1, observationCounter)
+
+        observationCounter = 0
```
You should at least mention that the throwing observer got unsubscribed when it threw - otherwise, it is completely unclear why it is "fixed" the second time round, at least to the casual reader who does not know the internal quirks of Rx.Observables.

blsemo commented on core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt:
```diff
@@ -212,32 +212,53 @@ class FinalityFlow private constructor(val transaction: SignedTransaction,
         }
     }
 
+
```
Superfluous blank lines

blsemo commented on core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt:
```diff
@@ -212,32 +212,53 @@ class FinalityFlow private constructor(val transaction: SignedTransaction,
     @Suspendable
     private fun notariseAndRecord(): SignedTransaction {
-        val notarised = if (needsNotarySignature(transaction)) {
+        val notaryCandidate = if (needsNotarySignature(transaction)) {
```
Why'd you rename `notarised` to `notaryCandidate`? I am not sure that makes it any more readable - also it's not a candidate for notarisation, it is either notarised or not.

kyriathar replied:
I mostly did it for cases like `hasNotarySignature(notarised)`. Ok, going back to `notarised`.

blsemo commented on core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt:
```diff
@@ -212,32 +212,53 @@ class FinalityFlow private constructor(val transaction: SignedTransaction,
+        } catch (e: Exception) {
+            throw if (hasNotarySignature(notaryCandidate)) {
+                logger.warn(
+                    "Caught an Exception for the notarised transaction ${notaryCandidate.id} when trying to record it locally. " +
```
I think this warning message is genuinely to long - should be a shorter message, and maybe refer to the documentation or similar where the problem is described in more detail.
Also, the wording is not ideal - I would write something like:
`Caught an exception while trying to record notarised transaction ${notaryCandidate.id} locally - the affected states are in an inconsistent state between the node and the notary. Hospitalising the flow.`

blsemo commented on core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt:
```diff
@@ -212,32 +212,53 @@ class FinalityFlow private constructor(val transaction: SignedTransaction,
+                            "meaning none the transaction nor any of its output states will be recorded in any vault. " +
+                            "However, re-consuming its input states with a new flow and a new transaction is no longer possible " +
+                            "since these states are now recorded as consumed within the Notary. " +
+                            "*** Therefore, we need not loose, by any means, the current transaction (${notaryCandidate.id}) *** " +
```
You want "lose" here - you can only "loose" an arrow or similar when you shoot it.

kyriathar-r3 committed:
Minor changes to test code.

blsemo reviewed (dismissed):
Generally ok, but the tests need some work to not become candidates for flakiness, and the warning message needs rewording.

rick-r3 reviewed (changes requested):
I don't think this is the correct approach.  We want any failure in `recordTransactions` to be trapped and handled.  `FinalityFlow` is too much of a special case.  I don't think we should necessarily bother with the special case of transactions that don't need notarising, and just treat all the failures the same.

Also, are the outcomes / tests not impacted by https://r3-cev.atlassian.net/browse/CORDA-3381 ?  It seems like maybe all the tests would need re-working if we prevent the unsubscribe ?  If so, maybe we should do that first?

kyriathar-r3 committed:
Changing the previous approach into wrapping all exceptions thrown from serviceHub#recordTransactions, in any case -and not only for FinalityFlow-, with a HospitalizeFlowException, throwing it instead.
Reverting FinalityFlow - leaving in the change on FinalityFlow#hasNotarySignature.
Tests were adjusted with the new approach and they were optimized using a Semaphore to signal flow hospitalization instead of wait for TimeoutException.

mnesbit commented on node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt:
```diff
@@ -393,12 +395,30 @@ class NodeVaultService(
+                            "" // any other exception type here is quite unexpected
+                        }
+                    }
+                    log.warn(
```
Why is this only at warn level? Sounds pretty ominous to me!

kyriathar-r3 committed:
Increasing logging level to ERROR.

kyriathar-r3 committed:
Minor text updates

kyriathar-r3 committed:
Change test of sql syntax error from observer to hit SedationNurse

kyriathar-r3 committed:
Make HospitalizeFlowException not suppress-able by user code

kyriathar-r3 committed:
Changed logic of test that would assert suppresion of exception thrown from observer by the flow, to hospitalising the flow.

kyriathar-r3 committed:
Changed bad test name and description

kyriathar-r3 committed:
Shortening code and renaming variable

kyriathar-r3 committed:
Removed redundant test which also had a misleading name in terms of 'GetsRetried'

kyriathar-r3 committed:
Minor code refactoring in exception catches

kyriathar-r3 committed:
Revert "Removed redundant test which also had a misleading name in terms of 'GetsRetried'"

This reverts commit 5feb1558

kyriathar-r3 committed:
Revert "Changed bad test name and description"

This reverts commit e86e18d9

kyriathar-r3 committed:
Fixing typos, minor changes to comments, adding TODOs to be fixed in later commits

kyriathar-r3 committed:
Remove no action TODO, test works as intended

kyriathar-r3 committed:
Remove never reachable code from DatabaseEndocrinologist.customConditions added lamdas, due to short-circuit AND in DatabaseEndocrinologist#consult

kyriathar-r3 committed:
Correcting test, persistence exception will occur during record transactions in NodeVaultService#processAndNotify

kyriathar-r3 committed:
Remove redundant test, persistence exception occurs during record transactions in NodeVaultService#processAndNotify, therefore never reaching vault observer code. Test 'persistenceExceptionDuringRecordTransactionsCannotBeSuppressedInFlow' already covers the current behaviour of this test

kyriathar-r3 committed:
Correct test description and name to reflect the current behaviour, code never reaches observer code

kyriathar-r3 committed:
Remove redundant test, persistence exception occurs during record trancastions in NodeVaultService#processAndNotify, therefore never reaching on commit. Test 'persistenceExceptionDuringRecordTransactionsGetsKeptForObservation' already covers the current behaviour of this test

kyriathar-r3 committed:
Modifying test, DbListenerService and CreateStateFlow to assert that none exception, thrown from vault observer, can be suppressed in flow

kyriathar-r3 committed:
Remove no action TODO

kyriathar-r3 committed:
Removing test because none exception thrown in vault observer can any longer bring a flow down. Flow will get hospitalised

kyriathar-r3 committed:
Revert FinalityFlowTests.kt to 14050826e92d4c93f4c34f166600c570a52b5ba6
- deleting those tests as they are already covered by tests in
VaultObserverExceptionTest

kyriathar-r3 committed:
Remove not making sense comment

kyriathar-r3 committed:
Line break

kyriathar-r3 committed:
Update Detekt baseline

lankydan commented on core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt:
```diff
@@ -231,13 +231,13 @@ class FinalityFlow private constructor(val transaction: SignedTransaction,
         val wtx = stx.tx
         val needsNotarisation = wtx.inputs.isNotEmpty() || wtx.references.isNotEmpty() || wtx.timeWindow != null
-        return needsNotarisation && hasNoNotarySignature(stx)
+        return needsNotarisation && !hasNotarySignature(stx)
```
I don't think there is any need to change this to `!hasNotarySignature` I think it was clear enough as it was.

kyriathar-r3 committed:
Changed the way we treat errors thrown in CordaPersistence#transaction. It would previously register an Event.Error in the fiber's channel, while now it brings the DatabaseTransaction
into a "corrupted" state. Because the Event.Error was subscribed in the fiber's queue it was not processed immediately
and therefore a checkpoint -which gets handled immediately- could precede the Event.Error handling.

kyriathar-r3 committed:
Revent back mistakenly changed behaviour in CordaPersistence#createSession

kyriathar-r3 committed:
Remove redundant test code, detekt updates

kyriathar-r3 committed:
Update Detekt baseline

kyriathar-r3 committed:
Bring back 'mentionsThrowable' method name

kyriathar-r3 committed:
Minor code format change

blsemo commented on node/src/main/kotlin/net/corda/node/services/statemachine/StaffedFlowHospital.kt:
```diff
@@ -595,15 +598,33 @@ class StaffedFlowHospital(private val flowMessaging: FlowMessaging,
+        true
+    } else {
+        atDepth == currDepth
+    }
```
If `atDepth` is not null, there is no point of further recursion, is there? In this case, we look for an exception at a specific depth in the chain, and if that doesn't match, we return false right away.

blsemo commented on node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt:
```diff
@@ -1244,14 +1243,17 @@ fun createCordaPersistence(databaseConfig: DatabaseConfig,
+            FlowStateMachineImpl.currentStateMachine()?.let {
+                // register the very first exception thrown throughout a chain of logical transactions
+                if (logicalTxCorruptedBy == null) logicalTxCorruptedBy = t
+            }
```
What do we need the flow state machine for here? Is that just to check that we are inside a flow?

blsemo commented on node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/CordaPersistence.kt:
```diff
@@ -230,18 +229,15 @@ class CordaPersistence(
-            errorHandler(persistenceException)
-            throw persistenceException
+        } else {
+            inTopLevelTransaction(isolationLevel, recoverableFailureTolerance, recoverAnyNestedSQLException, statement)
```
Might be a good idea to explain why top level transactions are treated differently here - my guess is because that call will not be able to suppress the exception anyway?

blsemo reviewed (commented):
If I understand the code correctly, scheduling of an error event is replaced by hooking the transaction and making sure that any exception that might have been suppressed is rethrown on commit, relying on the fact that the state machine will commit the transaction and send the flow to the flow hospital if that throws?
I think this would warrant at least a bit of a comment, as following this thought is not at all obvious.

kyriathar-r3 committed:
Adding explanatory comments

kyriathar-r3 committed:
Detekt update

mnesbit commented:
Can we review if there is enough test coverage. I had expected at least a few tests added, or else a change to a test to cover teh new behaviour.

kyriathar-r3 committed:
Revert FinalityFlow to ef57e9909c92fc18ad568969fde5483d3f8cf893 - changing hasNotarySignature back to hasNoNotarySignature

kyriathar-r3 committed:
Make recursion in mentionsThrowable jump fast

lankydan commented on node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/DatabaseTransaction.kt:
```diff
@@ -51,7 +50,12 @@ class DatabaseTransaction(
     private var closed = false
 
+    // is to be used when a logical transaction/ chain of logical transactions throws certain exceptions within a state machine
+    var logicalTxCorruptedBy: Throwable? = null // TODO: probably not needed to be set in case of SQLException or PersistenceException?
```
This `TODO` will need to be addressed

lankydan commented on node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/DatabaseTransaction.kt:
```diff
@@ -51,7 +50,12 @@ class DatabaseTransaction(
+
     fun commit() {
+        if (logicalTxCorruptedBy != null)
+            throw logicalTxCorruptedBy as Throwable
```
Does this need to be cast to `Throwable`? I thought kotlin should infer this type because of the null check above.

kyriathar replied:
Because it is a mutable property, some other thread could have modified this one to null between != null and throwing it.

lankydan commented on node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/DatabaseTransaction.kt:
```diff
@@ -51,7 +50,12 @@ class DatabaseTransaction(
     private var committed = false
     private var closed = false
 
+    // is to be used when a logical transaction/ chain of logical transactions throws certain exceptions within a state machine
```
Could we expand on this comment?

lankydan commented on node/src/integration-test/kotlin/net/corda/node/services/vault/VaultObserverExceptionTest.kt:
```diff
@@ -72,92 +66,51 @@ class VaultObserverExceptionTest {
-    @Test
-    fun persistenceExceptionOnCommitGetsRetriedAndThenGetsKeptForObservation() {
-        var admitted = 0
+    fun noneExceptionFromVaultObserverCanBeSuppressedInFlow() {
```
Rename the test `exceptionFromVaultObserverCannotBeSuppressedInFlow`

lankydan commented on node/src/integration-test/kotlin/net/corda/node/services/vault/VaultObserverExceptionTest.kt:
```diff
@@ -310,4 +219,86 @@ class VaultObserverExceptionTest {
+    }
+
+//    @Test
+//    fun `calling AppServiceHub#jdbcSession in observer code breaks transactional - leaves rubbish in the database - retry from checkpoint should find duplicates `() {
```
Remember to tidy this test up

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/StaffedFlowHospital.kt:
```diff
@@ -595,15 +598,36 @@ class StaffedFlowHospital(private val flowMessaging: FlowMessaging,
+    exceptionType: Class<T>,
+    errorMessage: String? = null,
+    atDepth: Int? = null,
+    currDepth: Int = 0
```
Can we just name this `currentDepth` ?

kyriathar replied:
Flow Hospital is reverted back; depths are removed from `mentionsThrowable` .

kyriathar-r3 committed:
Tidy up test code as per comments

kyriathar-r3 committed:
Minor code formattings

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt:
```diff
@@ -393,12 +395,23 @@ class NodeVaultService(
+                    // it could cause the ledger go into an inconsistent state, therefore we should hospitalise this flow
+                    // observer code should either be fixed or ignored and have the flow retry from previous checkpoint
+                    val causedBy = when (e) {
+                        is OnErrorNotImplementedException -> "- caused by an exception thrown in a rx.Observer#onNext that was unhandled (the observer has been unsubscribed!) "
```
in _an_ rx.Observer

(same in the lin below)

rick-r3 commented on testing/cordapps/dbfailure/dbfworkflows/src/main/kotlin/com/r3/dbfailure/workflows/DbListenerService.kt:
```diff
@@ -80,6 +80,10 @@ class DbListenerService(services: AppServiceHub) : SingletonSerializeAsToken() {
                             throw InvalidParameterException("Toys out of pram")
                         }
+                        CreateStateFlow.ErrorTarget.ServiceThrowMotherOfAllExceptions -> {
+                            log.info("Throw Exception")
```
Could you add 2 more tests, one for `OnErrorNotImplementedException` and one for `OnErrorFailedException` so we exercise all the code paths (even though they are trivial)?

kyriathar replied:
All the existing tests, throwing an exception in a vault observer, they are failing with a `OnErrorNotImplementedException`. 

Will add test that will fail with `OnErrorFailedException`.

kyriathar-r3 committed:
Not implementing TODO, we should just carry SQLException and PersistenceException to statemachine transitioning as well

kyriathar-r3 committed:
Replace ugly casting as per comment

kyriathar-r3 committed:
Add explanatory comment as per PR comment

kyriathar-r3 committed:
Add runtime exception test

kyriathar-r3 committed:
Update comment

kyriathar-r3 committed:
Variable renaming

kyriathar-r3 committed:
Added kdoc to test and updated its name

kyriathar-r3 committed:
Change var name and description

kyriathar-r3 committed:
Make DbListenerService accept onError definitions, hooked on companion object

kyriathar-r3 committed:
Add test, asserting OnErrorFailedException comming out of vault observer gets hospitalised

kyriathar-r3 committed:
Distinquish DatabaseTransactionException from StateTransitionException comming out of state machine transitioning. This way StaffedFlowHospital was reverted back

kyriathar-r3 committed:
Move comment

kyriathar-r3 committed:
Address detekt issue

kyriathar-r3 committed:
Unwrapping OnErrorNotImplementedException, throwing wrapped SQLException or PersistenceException plain

kyriathar-r3 committed:
Added test, asserting that on jvm halt, upon node restart flow completes successfully

kyriathar-r3 committed:
Adding comment

kyriathar-r3 committed:
Update Detekt baseline

lankydan commented on node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/DatabaseTransaction.kt:
```diff
@@ -71,6 +87,7 @@ class DatabaseTransaction(
     fun close() {
         if (sessionDelegate.isInitialized() && session.isOpen) {
             session.close()
+            firstExceptionInDatabaseTransaction = null
```
Does this work with the `set` function which null checks?

lankydan commented on node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/DatabaseTransaction.kt:
```diff
@@ -51,7 +52,22 @@ class DatabaseTransaction(
+     * The purpose of this property is to make sure this exception cannot be suppressed in user code.
+     * The exception will be thrown on the next [commit]. It is used only inside a flow state machine execution.
+     */
+    var firstExceptionInDatabaseTransaction: Throwable? = null
```
Should this be an `Exception` instead of a `Throwable`? I don't think we ever pass `Error`s to here, so it shouldn't matter too much but might be worth changing.

kyriathar-r3 committed:
Change DatabaseTransaction.firstExceptionInDatabaseTransaction type to Exception

lankydan commented on testing/cordapps/dbfailure/dbfworkflows/src/main/kotlin/com/r3/dbfailure/workflows/DbListenerService.kt:
```diff
@@ -12,10 +16,21 @@ class DbListenerService(services: AppServiceHub) : SingletonSerializeAsToken() {
         val log = contextLogger()
+        var onError: ((Throwable) -> Unit)? = null
+
+        // make flow executed on a outOfProcess node throw an unrecoverable error the first time only
```
No need to put the flow into the companion object

kyriathar-r3 committed:
Remove 'MakeServiceThrowErrorFlow' from companion object

kyriathar-r3 committed:
Proper nulling of DatabaseTransaction.firstExceptionInDatabaseTransaction

kyriathar-r3 committed:
Fixed: Nulling firstExceptionInDatabaseTransaction was not working

kyriathar-r3 committed:
Flush session before run block of code on EntityManager

rick-r3 reviewed (dismissed):
LGTM.  Let's try and get the integration tests to pass.

rick-r3 reviewed (dismissed):
Approving again after rebase

rick-r3 reviewed (approved):
Approve after rebase
